### PR TITLE
Prototype named actors.

### PR DESCRIPTION
**Pull request by heyucongtom**
## What do these changes do?

Since it's a total re-factorization, I feel like it may be cleaner to open up a new commit tree for the modification. It refer to the same issue as discussed in the pull-request #2120. (Please feel free to close that one.)

This commit deal with:
1. Implement the get_actor and register_actor inside the experimental directory.
2. Add unit-test for single driver environment.
3. Raise appropriate errors for the operation.

## Related issue number
#1424 #2120

### CODE REVIEW

File: test/actor_test.py [hunk 1]
```diff
@@ -1906,6 +1906,36 @@ def method(self):
...
+            def method(self):
+                pass
+
+        f1 = Foo.remote()
+        # Test saving f
+        ray.experimental.register_actor("f1", f1)
+        # Test getting f
+        f2 = ray.experimental.get_actor("f1")
+        self.assertEqual(f1._actor_id, f2._actor_id)
+
```
Comment (robertnishihara) [inline test/actor_test.py]: Let's modify the actor to be something like

```
def method(self):
    self.x += 1
    return self.x
```

so that here we can do

```
self.assertEqual(ray.get(f1.method.remote(), 1)
self.assertEqual(ray.get(f2.method.remote(), 2)
self.assertEqual(ray.get(f1.method.remote(), 3)
self.assertEqual(ray.get(f2.method.remote(), 4)
```

File: test/actor_test.py [hunk 2]
```diff
@@ -1906,6 +1906,36 @@ def method(self):
...
+        with self.assertRaises(AssertionError):
+            ray.experimental.register_actor("f1", f2)
+
+        # Test register with wrong object type
+        with self.assertRaises(AssertionError):
+            ray.experimental.register_actor("f3", 1)
+
+        # Test getting an unexist actor
+        with self.assertRaises(AssertionError):
+            err = ray.experimental.get_actor("unexisted")
```
Comment (robertnishihara) [inline test/actor_test.py]: unexisted -> nonexistent

File: test/actor_test.py [hunk 3]
```diff
@@ -1906,6 +1906,36 @@ def method(self):
         # we should also test this from a different driver.
         ray.get(new_f.method.remote())
 
+    def testRegisterAndGetActorHandle(self):
```
Comment (robertnishihara) [inline test/actor_test.py]: can we change this to `testRegisterAndGetNamedActors`? It'll be easier to find by searching

File: python/ray/experimental/named_actors.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+import ray
+import ray.cloudpickle as pickle
+"""
+This file contains functions intended to implement the named actor
+"""
+
+def _calculate_key_(name):
+    return b"Actor:" + str.encode(name)
+
+def get_actor(name):
```
Comment (robertnishihara) [inline python/ray/experimental/named_actors.py]: Please add docstrings for these methods with the format used e.g., in

https://github.com/ray-project/ray/blob/4584193308a749bdea6dd757fc0148c2950d0b30/python/ray/worker.py#L2443-L2458

File: python/ray/experimental/named_actors.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+def register_actor(name, actor_handle):
+    worker = ray.worker.get_global_worker()
+    actor_hash = _calculate_key_(name)
+    assert type(actor_handle) == ray.actor.ActorHandle, \
+        "Error: you could only store named-actors."
+    is_existed = worker.redis_client.hexists(actor_hash, name)
+    assert not is_existed, \
+        "Error: the actor with name={} already exists".format(name)
+    pickled_state = pickle.dumps(actor_handle)
+    worker.redis_client.hmset(actor_hash, {name: pickled_state})
```
Comment (robertnishihara) [inline python/ray/experimental/named_actors.py]: There is a race condition with this approach. It's possible that another worker could add publish a named actor with the same name between the calls to `hexists` and `hmset`. You can avoid this with `hsetnx`. See https://redis.io/commands/hsetnx.

File: python/ray/experimental/named_actors.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
```
Comment (robertnishihara) [inline python/ray/experimental/named_actors.py]: Let's reserve assertions for checking things that should never actually happen. The errors (in both `register_actor` and `get_actor`) can happen if the user passes in the wrong values, so we should raise some sort of exception

File: python/ray/experimental/named_actors.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+        "Error: Multiple actors under this name."
+    assert pickled_state[0] is not None, \
+        "Error: actor with name {} doesn't exist".format(name)
+    handle = pickle.loads(pickled_state[0])
+    return handle
+
+def register_actor(name, actor_handle):
+    worker = ray.worker.get_global_worker()
+    actor_hash = _calculate_key_(name)
+    assert type(actor_handle) == ray.actor.ActorHandle, \
```
Comment (robertnishihara) [inline python/ray/experimental/named_actors.py]: should raise `TypeError`

File: python/ray/experimental/named_actors.py [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+        "Error: you could only store named-actors."
+    is_existed = worker.redis_client.hexists(actor_hash, name)
+    assert not is_existed, \
```
Comment (robertnishihara) [inline python/ray/experimental/named_actors.py]: should raise `ValueError`

File: python/ray/experimental/named_actors.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+This file contains functions intended to implement the named actor
+"""
+
+def _calculate_key_(name):
+    return b"Actor:" + str.encode(name)
+
+def get_actor(name):
+    worker = ray.worker.get_global_worker()
+    actor_hash = _calculate_key_(name)
+    pickled_state = worker.redis_client.hmget(actor_hash, name)
```
Comment (robertnishihara) [inline python/ray/experimental/named_actors.py]: You can just use `hget` and then remove the assertion below.

File: python/ray/experimental/named_actors.py [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+    assert len(pickled_state) == 1, \
+        "Error: Multiple actors under this name."
+    assert pickled_state[0] is not None, \
```
Comment (robertnishihara) [inline python/ray/experimental/named_actors.py]: should raise `ValueError`

**Review by robertnishihara — COMMENTED** [review]
This looks great! Please make sure the linting passes.

**Comment by robertnishihara** [comment]
To do linting, you can do (from `ray/`)

```
yapf --style .style.yapf -i  python/ray/experimental/__init__.py  python/ray/experimental/named_actors.py test/actor_test.py
```

also

```
flake8 python/ray/experimental/__init__.py  python/ray/experimental/named_actors.py test/actor_test.py
```

File: python/ray/experimental/__init__.py [hunk 11]
```diff
@@ -7,10 +7,13 @@
     flush_redis_unsafe, flush_task_and_object_metadata_unsafe,
     flush_finished_tasks_unsafe, flush_evicted_objects_unsafe,
     _flush_finished_tasks_unsafe_shard, _flush_evicted_objects_unsafe_shard)
+from .named_actors import (
+    get_actor, register_actor
+)
```
Comment (robertnishihara) [inline python/ray/experimental/__init__.py]: looks like this can all fit on one line without the parentheses

File: python/ray/experimental/named_actors.py (hunk 6 above)
Comment (heyucongtom) [reply python/ray/experimental/named_actors.py]: Handled.

File: test/actor_test.py (hunk 1 above)
Comment (heyucongtom) [reply test/actor_test.py]: Handled.

File: python/ray/actor.py [hunk 12]
```diff
@@ -869,7 +869,8 @@ def _serialization_helper(self, ray_forking):
             _ray_actor_creation_dummy_object_id.id(),
             "actor_method_cpus": self._ray_actor_method_cpus,
             "actor_driver_id": self._ray_actor_driver_id.id(),
-            "previous_actor_handle_id": self._ray_actor_handle_id.id(),
+            "previous_actor_handle_id": self._ray_actor_handle_id.id()
+            if self._ray_actor_handle_id else None,
```
Comment (robertnishihara) [inline python/ray/actor.py]: Why is this change needed?

Comment (robertnishihara) [reply python/ray/actor.py]: Hmm.. ok I see that the test failures without it.

**Review by robertnishihara — APPROVED** [review]
Thanks! I pushed some stylistic changes. I'll merge this once the tests pass.

**Review by robertnishihara — APPROVED** [review]

**Comment by robertnishihara** [comment]
One thing I noticed is that calling `get_actor` twice on the same name will cause problems (once you try to use the actor). This is probably the same issue as https://github.com/ray-project/ray/issues/2115.

**Comment by heyucongtom** [comment]
I just came across this bug thread earlier today. Would take a look at it
later.
